preprocessing/__init__.py

```diff
@@ -81,6 +82,84 @@ def check_complement(g: Graph, h: Graph) -> (Graph, Graph):
         return g, h
 
 
+def find_components(g: Graph):
+    """
+    Breadth First Search Alg. Which also:
+    tests if the graph is connected,
+    computes the distance from s to the other edges
+    labels the vertices in the order they are visited
+    :param g: Graph
+    :return: (isConnected, {dict of components})
+    """
+    visited = set()
+    components = dict()
+    count = 1
+
+    while len(visited) < len(g.vertices):
+        for o in g.vertices:
+            if o not in visited:
+                v = o
+                break
+        queue = DoubleLinkedList()  # queue
+        queue.append(v)
+        visited.add(v)
+        components[count] = [v]
+
+        while len(queue) > 0:
+            w = queue.pop()  # BFS so FIFO
+            for n in w.neighbours:
+                # If w not visited
+                if n not in visited:
+                    visited.add(n)
+                    queue.append(n)
+                    components[count].append(n)
+        count += 1
+
+    is_connected = len(components) == 1
+    return is_connected, components
+
+
+def construct_graph_from_components(components: dict) -> [Graph]:
+    """
+    constructs a list of graphs from a dictionary of components
+
+        :param components: a dictionary constructed from an (unconnected) graph
+        :return: list of subgraphs
+        """
+    graphs = list()
+    for key in components.keys():
+        subgraph = Graph(False)
+        for vertex in components[key]:
+            if vertex not in subgraph.vertices:
+                subgraph.add_vertex(vertex)
```

Comment:
Volgens mij is deze check niet nodig omdat je de subgraph nieuw aanmaakt en over de vertices loopt en dat de enige plek is waar je een vertex toevoegt. Dus dan is ieder vertex uniek - AG